### PR TITLE
Performance optimizations

### PR DESCRIPTION
**Pull request by robertlong**
- Combines gatherComponentsData and findDirtyComponents to remove duplicate work
- Instead of using CSS selectors every update in gatherComponentsData and findDirtyComponents we cache the results and use those if they exist.
- Moves the combined gatherComponentsData function into the networked component to store the cached elements.
- Gets rid of "child schema keys" and instead uses the index of the "component schema" to identity networked component data. This reduces a lot of work concatenating and comparing strings as well as reduces overall bandwidth due to the smaller identifier.
- Uses [fast-deep-equal](https://github.com/epoberezkin/fast-deep-equal) because apparently it's "The fastest deep equality check" by some benchmark.
- Removes an object allocation in createSyncData
- ~~Removes an allocation in isFullSync due to the usage of Object.keys() which allocates an array.~~
- Includes @johnshaughnessy's `isFirstSync` PR: https://github.com/MozillaReality/networked-aframe/pull/8

Todo:
- [x] Look for more memory allocations that occur every tick or network update.
- [ ] Investigate a faster alternative to cloning component data.
- [x] Fix tests.

A future PR I'd like to make is to switch to sending a single network update message instead of multiple in a single frame and sometimes multiple in back to back frames. I'm not certain this will help, but I have a hunch that it will have some effect on performance by batching WebRTC/Websocket messages.

### CODE REVIEW

**Comment by mqp** [comment]
I didn't read this perfectly carefully, but I didn't spot any bugs yet :+1:

**Comment by mqp** [comment]
> Investigate a faster alternative to cloning component data.

Correct me if I'm wrong, but in the code as written, it looks like we clone `syncedComponentData` and replace the entry in `cachedData` even if they were already deep-equal, as we determined just above. Should we skip the replacement in that case?

**Comment by haydenjameslee** [comment]
I'd be interested in seeing the comparison between deep equal functions. We care about speed but also memory allocation. I'm not sure how to resolve the tradeoff between the two.

I'll also hold off on merging this branch: https://github.com/networked-aframe/networked-aframe/pull/121

File: src/NetworkEntities.js [hunk 1]
```diff
@@ -59,18 +61,22 @@ class NetworkEntities {
...
+    } else if (this.isFullSync(entityData)) {
       this.receiveFirstUpdateFromEntity(entityData);
     }
   }
 
   isFullSync(entityData) {
-    var numSentComps = Object.keys(entityData.components).length;
+    var numSentComps = 0;
+    // eslint-disable-next-line no-unused-vars
+    for (var componentIndex in entityData.components) {
```
Comment (netpro2k) [inline src/NetworkEntities.js]: John had another PR to just tag "full syncs" so we dont have to do this weird guess at all.

File: src/components/networked.js [hunk 2]
```diff
@@ -83,8 +92,16 @@ AFRAME.registerComponent('networked', {
...
       this.el.setAttribute('networked', { owner: NAF.clientId });
       this.syncAll();
-      this.el.emit(this.OWNERSHIP_GAINED, { el: this.el, oldOwner: owner });
-      this.el.emit(this.OWNERSHIP_CHANGED, { el: this.el, oldOwner: owner, newOwner: NAF.clientId});
+
+      this.onOwnershipGainedEvent.el = this.el;
+      this.onOwnershipGainedEvent.oldOwner = owner;
+      this.el.emit(this.OWNERSHIP_GAINED, this.onOwnershipGainedEvent);
+
+      this.onOwnershipChangedEvent.el = this.el;
```
Comment (netpro2k) [inline src/components/networked.js]: This doesn't change, might as well initialize that above when you create `ownershipChangedEvent`

File: src/components/networked.js [hunk 3]
```diff
@@ -308,62 +368,55 @@ AFRAME.registerComponent('networked', {
       const oldOwner = this.data.owner;
       const newOwner = entityData.owner;
       if (wasMine) {
-        this.el.emit(this.OWNERSHIP_LOST, { el: this.el, newOwner: newOwner });
+        this.onOwnershipLostEvent.el = this.el;
```
Comment (netpro2k) [inline src/components/networked.js]: same here, just set in the constructor

File: src/components/networked.js [hunk 4]
```diff
@@ -308,62 +368,55 @@ AFRAME.registerComponent('networked', {
...
+        this.onOwnershipLostEvent.newOwner = newOwner;
+        this.el.emit(this.OWNERSHIP_LOST, this.onOwnershipLostEvent);
       }
-      this.el.emit(this.OWNERSHIP_CHANGED, { el: this.el, oldOwner: oldOwner, newOwner: newOwner});
+      this.onOwnershipChangedEvent.el = this.el;
```
Comment (netpro2k) [inline src/components/networked.js]: And again here, set in the constructor

**Review by netpro2k — COMMENTED** [review]
Would still like to do a second pass when you are done but so far looking good!

File: src/NetworkEntities.js (hunk 1 above)
Comment (haydenjameslee) [reply src/NetworkEntities.js]: +1

File: src/NetworkEntities.js [hunk 5]
```diff
@@ -137,12 +133,7 @@ class NetworkEntities {
   completeSync(targetClientId) {
     for (var id in this.entities) {
       if (this.entities.hasOwnProperty(id)) {
-        this.onSyncAllEvent.targetClientId = targetClientId;
-        this.entities[id].emit(
-          'syncAll',
-          this.onSyncAllEvent,
-          false
-        );
+        this.entities[id].components.networked.syncAll(targetClientId);
```
Comment (netpro2k) [inline src/NetworkEntities.js]: :tada:

File: src/NetworkEntities.js [hunk 6]
```diff
@@ -62,21 +62,11 @@ class NetworkEntities {
 
     if (this.hasEntity(networkId)) {
       this.entities[networkId].components.networked.networkUpdate(entityData);
-    } else if (this.isFullSync(entityData)) {
+    } else if (entityData.isFirstSync) {
```
Comment (netpro2k) [inline src/NetworkEntities.js]: I think this should be `isFullSync` not `isFirstSync` I think there are other cases we may want to send full syncs, such as after ownership transfer.

File: src/components/networked.js [hunk 7]
```diff
@@ -136,7 +136,7 @@ AFRAME.registerComponent('networked', {
           NAF.log.warn("Networked element was removed before ever getting the chance to syncAll");
           return;
         }
-        this.syncAll();
+        this.syncAll(undefined, true);
```
Comment (netpro2k) [inline src/components/networked.js:159]: Hmm this feels fishy. Should this just be `this.updateNextSyncTime()`? That appears to be the only thing actually run when `targetClientId` is falsey.

File: src/components/networked.js [hunk 8]
```diff
@@ -197,7 +197,7 @@ AFRAME.registerComponent('networked', {
 
     var components = this.gatherComponentsData(true);
 
-    var syncData = this.createSyncData(components);
+    var syncData = this.createSyncData(components, isFirstSync);
 
     if (targetClientId) {
```
Comment (netpro2k) [inline src/components/networked.js]: Are there places other than onConenct where `syncAll` gets called without a `targetClientId`?

File: src/components/networked.js (hunk 7 above)
Comment (robertlong) [reply src/components/networked.js:159]: Isn't it also broadcasting the initial component data to all users on connection?

File: src/components/networked.js [hunk 9]
```diff
@@ -54,13 +54,7 @@ AFRAME.registerComponent('networked', {
...
     this.componentSchemas =  NAF.schemas.getComponents(this.data.template);
     this.cachedElements = new Array(this.componentSchemas.length);
-    this.networkUpdatePredicates = this.componentSchemas.map((componentSchema) => {
-      if (componentSchema.requiresNetworkUpdate) {
-        return componentSchema.requiresNetworkUpdate();
-      }
-
-      return defaultNetworkUpdatePredicate();
-    });
+    this.networkUpdatePredicates = this.componentSchemas.map(x => x.requiresNetworkUpdate || defaultRequiresUpdate());
```
Comment (robertlong) [inline src/components/networked.js]: Doesn't requireNetworkUpdate need to be called?

Ex:
```
this.networkUpdatePredicates = this.componentSchemas.map(x => x.requiresNetworkUpdate ? x.requiresNetworkUpdate() : defaultRequiresUpdate());
```

File: src/components/networked.js (hunk 7 above)
Comment (johnshaughnessy) [reply src/components/networked.js:159]: Yes, it seems that way https://github.com/MozillaReality/networked-aframe/pull/12/files#diff-69b32c2392de94b26df18f638d9c3053R208

File: src/components/networked.js (hunk 9 above)
Comment (robertlong) [reply src/components/networked.js]: Nvm, I misunderstood the value of `requiresNetworkUpdate`

**Pull request merged**